### PR TITLE
Scene view

### PR DESCRIPTION
**Pull request by nikitavbv**

### CODE REVIEW

**Comment by nikitavbv** [comment]
[Camera controls & scenes](https://trello.com/c/6PKiHhuI/4-camera-controls-scenes)

**Comment by nikitavbv** [comment]
[Camera controls & scenes](https://trello.com/c/6PKiHhuI/4-camera-controls-scenes)

File: src/main/java/labs/introtoprogramming/lab5/gui/Input.java [hunk 1]
```diff
@@ -0,0 +1,59 @@
+package labs.introtoprogramming.lab5.gui;
+
+import labs.introtoprogramming.lab5.geometry.Vector3;
+
+public class Input {
+
+  private boolean forwardKeyDown = false;
```
Comment (Brookoko) [inline src/main/java/labs/introtoprogramming/lab5/gui/Input.java:7]: Assignment is redundant. Is it made for clarity?

File: src/main/java/labs/introtoprogramming/lab5/scene/ControllableCamera.java [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+  private double movementSpeed = DEFAULT_MOVEMENT_SPEED;
+  private double mouseSensitivity = DEFAULT_MOUSE_SENSITIVITY;
+
+  public ControllableCamera(Raster raster, Transform transform) {
+    super(raster, transform);
+  }
+
+  @Override
+  protected void update(int delta) {
+    getTransform().setPosition(
```
Comment (Brookoko) [inline src/main/java/labs/introtoprogramming/lab5/scene/ControllableCamera.java]: Maybe is it better to use directly transform field in base class?

File: src/main/java/labs/introtoprogramming/lab5/scene/Scene.java [hunk 3]
```diff
@@ -1,17 +1,25 @@
...
+  public abstract List<PointLight> getLights();
 
-  Color getBackgroundColor();
+  public abstract Optional<Camera> getCamera();
 
+  abstract Color getBackgroundColor();
+
+  public abstract void addSceneObject(SceneObject obj);
+
+  public void addSceneObjects(SceneObject... obj) {
```
Comment (Brookoko) [inline src/main/java/labs/introtoprogramming/lab5/scene/Scene.java:26]: Maybe make this function abstract and make abstract class as interface?

File: src/main/java/labs/introtoprogramming/lab5/scene/SceneObject.java [hunk 4]
```diff
@@ -37,4 +41,12 @@ public Mesh getMesh() {
   public boolean intersect(Ray ray) {
     return false;
   }
+
+  public void setInput(Input input) {
```
Comment (Brookoko) [inline src/main/java/labs/introtoprogramming/lab5/scene/SceneObject.java]: Most of object are static and this functionality is redundant for them. Maybe is it more reasonable to get data from input and than apply it to object via transform?

File: src/main/java/labs/introtoprogramming/lab5/gui/SceneRendererWindow.java [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+  private static final int UPDATE_RATE = 1000/TARGET_FPS;
+  private static final boolean DEFAULT_CAMERA_CONTROLLABLE = false;
+
+  private SceneRender render;
+  private Scene scene;
+  private MouseController mouseController;
+  private Input input;
+
+  public SceneRendererWindow(Scene scene) {
+    this(new BasicRaytracingRender(scene.getCamera().orElseThrow(NoCameraException::new).raster()), scene);
```
Comment (Brookoko) [inline src/main/java/labs/introtoprogramming/lab5/gui/SceneRendererWindow.java]: Maybe is it better to refactor SceneRender to only accept scene, if raster can be obtain from camera now? Or is it out of place for this PR?

File: src/main/java/labs/introtoprogramming/lab5/gui/Input.java (hunk 1 above)
Comment (nikitavbv) [reply src/main/java/labs/introtoprogramming/lab5/gui/Input.java:7]: Yes, to make it clear that my default all keys are not down.

File: src/main/java/labs/introtoprogramming/lab5/scene/Scene.java (hunk 3 above)
Comment (nikitavbv) [reply src/main/java/labs/introtoprogramming/lab5/scene/Scene.java:26]: This method implementation is likely to stay the same for all implementations. So doesn't make a lot of scene to make it abstract.

File: src/main/java/labs/introtoprogramming/lab5/gui/SceneRendererWindow.java (hunk 5 above)
Comment (nikitavbv) [reply src/main/java/labs/introtoprogramming/lab5/gui/SceneRendererWindow.java]: Multiple constructors increase flexibility. There should be a way to set custom scene render.

File: src/main/java/labs/introtoprogramming/lab5/gui/Input.java (hunk 1 above)
Comment (nikitavbv) [reply src/main/java/labs/introtoprogramming/lab5/gui/Input.java]: Yes, to make it clear that all keys are not down by default.

File: src/main/java/labs/introtoprogramming/lab5/scene/Scene.java (hunk 3 above)
Comment (nikitavbv) [reply src/main/java/labs/introtoprogramming/lab5/scene/Scene.java]: Most implementations are likely to have the same logic for this method, so it is better to keep it as it is.

File: src/main/java/labs/introtoprogramming/lab5/gui/SceneRendererWindow.java (hunk 5 above)
Comment (nikitavbv) [reply src/main/java/labs/introtoprogramming/lab5/gui/SceneRendererWindow.java]: Other constructors allow more flexibility, for example to set custom scene render.

Comment (Brookoko) [reply src/main/java/labs/introtoprogramming/lab5/gui/SceneRendererWindow.java]: I am talking about basicRaytracingRender. Maybe is it better to add constructor that only accept scene as argument and take raster from camera?

Comment (nikitavbv) [reply src/main/java/labs/introtoprogramming/lab5/gui/SceneRendererWindow.java]: Oh, right then.

**Review by Brookoko — APPROVED** [review]

**Pull request merged**